library/step.go

```diff
@@ -505,8 +506,35 @@ func (s *Step) String() string {
 	)
 }
 
+// InitFrom initializes some Step type fields based on a Container and a Build.
+func (s *Step) InitFrom(ctn *pipeline.Container, build *Build) {
```

Comment:
@jbrockopp any feedback on that ^?

Reply:
Apologies it took awhile for me to respond on this 😞 

> I avoided renaming the existing function so that PRs are easier to coordinate across repos.

Sure 👍 

I think that's a fair point and something to consider moving forward.

FWIW, we have made some efforts in an attempt to reduce the amount of repos.

I bring it up in case you have interest in renaming the existing function.

References:

* https://github.com/go-vela/community/issues/394
* https://github.com/go-vela/community/issues/395

> Another issue is that it's not just from the container and `StepFromContainerAndBuild` seemed a bit too wordy.

Yeah I could see that being a cause for concern.. one idea I have on that is to shorten the function name?

Perhaps it could be something like `StepFromBuildContainer`? or `StepFromContAndBuild`?

> What about `NewStepFrom`?

I think that could be something else to consider... although I feel like `New` in the name may be duplicative.

And I worry that `NewStepFrom` doesn't offer enough description for the function.

Reply:
I figured I'd use `New` in the name because it is not just converting between types, it also initializes the status field to  pending.

Now that the restructure dust has settled, I'll rename the existing function to `StepFromContainerEnvironment` since that is more accurate. Naming the new function is less clear.

Is it less clear because the function is doing too much? Maybe I should turn it into 2 Step receiver functions, FromBuild and FromContainer?

Reply:
Creating 2 receiver functions would work 👍 

Thinking through that approach.. Do you think we'd have to create another function that would merge them?

If so, I might have concerns around the extra complexity and LOC introduced by having multiple functions.

I'm likely not doing a good job of explaining my train of thought so I'll try to demonstrate with an example.

If we had the two functions separate, a code sample of that might look like:

```go
	// pretend these have data already set
	b := new(library.Build)
	ctn := new(pipeline.Container)

	// this is the object we want to convert from the above 'b'
	bStep := StepFromBuild(b)

	// this is the object we want to convert from the above 'ctn'
	ctnStep := StepFromContainer(ctn)

	// add code to handle merging the two objects together?
```

Perhaps my sample is off, but it almost seems like we'd need a function to merge `bStep` and `ctnStep`?

Reply:
Hmm. I was thinking something more like this:
```go
	// pretend these have data already set
	b := new(library.Build)
	ctn := new(pipeline.Container)

	// create a new step
	step := new(library.Step)

	// copy details from build to step
	step.CopyFromBuild(b)

	// copy details from container to step
	step.CopyFromContainer(ctn)
```

Or maybe we need to take the opts approach:
```go
	// pretend these have data already set
	b := new(library.Build)
	ctn := new(pipeline.Container)

	// create a new step
	step := NewStep(
		WithBuild(b),        // or maybe FromBuild(b)
		WithContainer(ctn),  // or maybe FromContainer(ctn)
	)
```
In this case, these would be something like this:
```go
func New(opts ...StepOpt) (*Step, error) {
	s := new(Step)

	// apply all provided configuration options
	for _, opt := range opts {
		err := opt(s)
		if err != nil {
			return nil, err
		}
	}

	// default status to Pending
	s.SetStatus(constants.StatusPending)

	return s, nil
}
func WithBuild(build library.Build) StepOpt {
	return func(s *Step) error {
		s.SetHost(build.GetHost())
		s.SetRuntime(build.GetRuntime())
		s.SetDistribution(build.GetDistribution())

		return nil
	}
}
func WithContainer(container pipeline.Container) StepOpt {
	return func(s *Step) error {
		s.SetName(ctn.Name)
		s.SetNumber(ctn.Number)
		s.SetImage(ctn.Image)

		// check if the VELA_STEP_STAGE environment variable exists
		value, ok := ctn.Environment["VELA_STEP_STAGE"]
		if ok {
			// set the Stage field to the value from environment variable
			s.SetStage(value)
		}

		return nil
	}
}
```

Reply:
Ahhh I see.. The only concerns I have with that approach is the increased level of complexity.

Meaning having to call multiple functions or learning what the Opts functions specifically do in the context.

I wonder if we're overcomplicating the current task we're trying to accomplish for this?

I feel like we should revisit some of your earlier suggestions around a single, long named function?

To be more specific, `StepFromContainerAndBuild` is `25` characters.

If we're willing to accept `StepFromContainerEnvironment` (`28` chars) then perhaps that's fine?

And if it's the function name itself, you could do a slight permutation of that, such as:

* `StepFromBuildAndContainer`
* `StepFromBuildContainer`

Reply:
OK. I went with `StepFromBuildContainer` wdyt?

Reply:
I think that should work great 👍